File: examples/ruby/test/hmac_signature_test.rb
require_relative 'test_helper'
require 'openssl'
require 'base64'

class HmacSignatureTest < Minitest::Test
  include Rack::Test::Methods

  SECRET = 'shhh'.freeze

  def app
    LocoAwareWebhook
  end

  def setup
    ENV['LOCOAWARE_WEBHOOK_SECRET'] = SECRET
    QueueClient::QUEUES.clear
  end

  def teardown
    ENV.delete('LOCOAWARE_WEBHOOK_SECRET')
  end

  def sign(body, secret)
    Base64.strict_encode64(OpenSSL::HMAC.digest('SHA256', secret, body))
  end

  def test_rejects_request_without_signature_header
    post '/webhooks/shipments', '{"shipment":{"id":"x"}}', { 'CONTENT_TYPE' => 'application/json' }
    assert_equal 401, last_response.status
  end

  def test_rejects_request_with_bad_signature
    body = '{"shipment":{"id":"x"}}'
    post '/webhooks/shipments', body, {
      'CONTENT_TYPE' => 'application/json',
      'HTTP_X_LOCOAWARE_SIGNATURE' => sign(body, 'wrong-secret'),
    }
    assert_equal 401, last_response.status
  end

  def test_accepts_correctly_signed_request
    body = '{"id":"evt_signed","shipment":{"id":"ship_signed"}}'
    Shipments.stub(:append_event, ->(_id, _evt) {}) do
      post '/webhooks/shipments', body, {
        'CONTENT_TYPE' => 'application/json',
        'HTTP_X_LOCOAWARE_SIGNATURE' => sign(body, SECRET),
      }
    end
    assert_equal 200, last_response.status
  end

  def test_accepts_correctly_signed_device_report
    body = '{"device":{"id":"dev_signed"}}'
    post '/webhooks/device-reports', body, {
      'CONTENT_TYPE' => 'application/json',
      'HTTP_X_LOCOAWARE_SIGNATURE' => sign(body, SECRET),
    }
    assert_equal 202, last_response.status
  end
end

File: examples/php/tests/HmacSignatureTest.php
<?php

declare(strict_types=1);

namespace LocoAware\Webhook\Tests;

use PHPUnit\Framework\TestCase;
use Slim\Psr7\Factory\ServerRequestFactory;
use Slim\Psr7\Factory\StreamFactory;
use function LocoAware\Webhook\createApp;

require_once __DIR__ . '/Doubles.php';

class HmacSignatureTest extends TestCase
{
    private const SECRET = 'shhh';

    private RecordingQueueClient $queue;
    private RecordingShipments $shipments;
    private RecordingDevices $devices;
    private RecordingAssets $assets;

    protected function setUp(): void
    {
        putenv('LOCOAWARE_WEBHOOK_SECRET=' . self::SECRET);
        $this->queue     = new RecordingQueueClient();
        $this->shipments = new RecordingShipments();
        $this->devices   = new RecordingDevices();
        $this->assets    = new RecordingAssets();
    }

    protected function tearDown(): void
    {
        putenv('LOCOAWARE_WEBHOOK_SECRET');
    }

    private function sign(string $body, string $secret): string
    {
        return base64_encode(hash_hmac('sha256', $body, $secret, true));
    }

    private function dispatch(string $body, array $headers = []): \Psr\Http\Message\ResponseInterface
    {
        $app = createApp($this->queue, $this->shipments, $this->devices, $this->assets);
        $request = (new ServerRequestFactory())->createServerRequest('POST', '/webhooks/shipments')
            ->withHeader('Content-Type', 'application/json')
            ->withBody((new StreamFactory())->createStream($body));
        foreach ($headers as $name => $value) {
            $request = $request->withHeader($name, $value);
        }
        return $app->handle($request);
    }

    public function testRejectsRequestWithoutSignatureHeader(): void
    {
        $resp = $this->dispatch(json_encode(['shipment' => ['id' => 'x']]));
        $this->assertSame(401, $resp->getStatusCode());
    }

    public function testRejectsRequestWithBadSignature(): void
    {
        $body = json_encode(['shipment' => ['id' => 'x']]);
        $resp = $this->dispatch($body, [
            'X-LocoAware-Signature' => $this->sign($body, 'wrong-secret'),
        ]);
        $this->assertSame(401, $resp->getStatusCode());
    }

    public function testAcceptsCorrectlySignedRequest(): void
    {
        $body = json_encode(['id' => 'evt_signed', 'shipment' => ['id' => 'ship_signed']]);
        $resp = $this->dispatch($body, [
            'X-LocoAware-Signature' => $this->sign($body, self::SECRET),
        ]);
        $this->assertSame(200, $resp->getStatusCode());
    }
}
